File: awatif-fem/src/cpp/utils/getTransformationMatrix.cpp
#include "../data-model.h"
#include <vector>
#include <cmath>
#include <Eigen/Dense>
#include <iostream>
#include <iomanip>
#include <stdexcept>

// Declarations
Eigen::MatrixXd getTransformationMatrixFrame(const Node &n0, const Node &n1);
Eigen::MatrixXd getTransformationMatrixShell(const Node &n0, const Node &n1, const Node &n2);

// Shell Q4 transformation — defined in shellQ4.cpp
Eigen::MatrixXd getTransformationMatrixShellQ4(
    const Node &n0, const Node &n1, const Node &n2, const Node &n3);

// Forward: check if 4-node element is shell Q4 (has thickness)
// We need elementInputs and index to decide, but the transformation dispatch
// doesn't have them. So for 4-node elements, we check if nodes are coplanar
// and non-coincident to decide shell vs interface.
// Simple heuristic: interface elements have coincident node pairs (0≈3, 1≈2).
static bool isShellQ4(const std::vector<Node> &nodes)
{
    // Interface: node 0 ≈ node 3, node 1 ≈ node 2 (zero-thickness)
    double d03 = std::sqrt(
        std::pow(nodes[0][0]-nodes[3][0], 2) +
        std::pow(nodes[0][1]-nodes[3][1], 2) +
        std::pow(nodes[0][2]-nodes[3][2], 2));
    double d12 = std::sqrt(
        std::pow(nodes[1][0]-nodes[2][0], 2) +
        std::pow(nodes[1][1]-nodes[2][1], 2) +
        std::pow(nodes[1][2]-nodes[2][2], 2));
    // If both pairs are nearly coincident → interface (zero-thickness)
    if (d03 < 1e-6 && d12 < 1e-6) return false;
    return true; // Otherwise → shell Q4
}

// Main dispatch function based on number of nodes
Eigen::MatrixXd getTransformationMatrix(const std::vector<Node> &nodes)
{
    if (nodes.size() == 2)
    {
        return getTransformationMatrixFrame(nodes[0], nodes[1]);
    }

    if (nodes.size() == 3)
    {
        return getTransformationMatrixShell(nodes[0], nodes[1], nodes[2]);
    }

    if (nodes.size() == 4)
    {
        if (isShellQ4(nodes))
        {
            return getTransformationMatrixShellQ4(nodes[0], nodes[1], nodes[2], nodes[3]);
        }
        // Interface element: K is computed directly in global coordinates
        return Eigen::MatrixXd::Identity(24, 24);
    }

    throw std::runtime_error("Unsupported element type in getTransformationMatrix (must have 2, 3, or 4 nodes).");
}

// Transformation matrix for 2-node frame/truss elements
Eigen::MatrixXd getTransformationMatrixFrame(const Node &n0, const Node &n1)
{
    Eigen::Vector3d vector(n1[0] - n0[0], n1[1] - n0[1], n1[2] - n0[2]);
    double length = vector.norm();
    if (length < 1e-12)
    {
        std::cerr << "Warning: Zero length element detected in getTransformationMatrixFrame. Returning 12x12 zero matrix." << std::endl;
        return Eigen::MatrixXd::Zero(12, 12);
    }

    double l = vector.x() / length;
    double m = vector.y() / length;
    double n = vector.z() / length;
    double D = std::sqrt(l * l + m * m);

    Eigen::Matrix3d lambda; // Direction cosine matrix

    if (std::abs(n - 1.0) < 1e-9) // Corresponds to n === 1 in TS
    {
        lambda << 0, 0, 1,
            0, 1, 0,
            -1, 0, 0;
    }
    else if (std::abs(n + 1.0) < 1e-9) // Corresponds to n === -1 in TS
    {
        lambda << 0, 0, -1,
            0, 1, 0,
            1, 0, 0;
    }
    else
    { // General case
        lambda << l, m, n,
            -m / D, l / D, 0,
            (-l * n) / D, (-m * n) / D, D;
    }

    // Construct the 12x12 transformation matrix T
    Eigen::MatrixXd T = Eigen::MatrixXd::Zero(12, 12);
    for (int i = 0; i < 4; ++i)
    {
        T.block<3, 3>(i * 3, i * 3) = lambda;
    }

    return T;
}

// Transformation matrix for 3-node shell elements
Eigen::MatrixXd getTransformationMatrixShell(
    const Node &n0_node,
    const Node &n1_node,
    const Node &n2_node)
{
    const int nodesCount = 3;
    const int dimensions = 3;
    const int dof = 6;
    const int totalDof = nodesCount * dof;

    Eigen::MatrixXd localCoordinates = Eigen::MatrixXd::Zero(dimensions, nodesCount);
    localCoordinates.col(0) << n0_node[0], n0_node[1], n0_node[2];
    localCoordinates.col(1) << n1_node[0], n1_node[1], n1_node[2];
    localCoordinates.col(2) << n2_node[0], n2_node[1], n2_node[2];

    const Eigen::Vector3d dN_dxi_ts(-1.0, 1.0, 0.0);
    const Eigen::Vector3d dN_det_ts(-1.0, 0.0, 1.0);

    Eigen::MatrixXd jacobianMatrix = Eigen::MatrixXd::Zero(dimensions, 2);

    for (int i = 0; i < dimensions; ++i)
    {
        for (int k = 0; k < nodesCount; ++k)
        {
            jacobianMatrix(i, 0) += localCoordinates(i, k) * dN_dxi_ts(k);
            jacobianMatrix(i, 1) += localCoordinates(i, k) * dN_det_ts(k);
        }
    }

    Eigen::Vector3d dX_dxi(jacobianMatrix(0, 0), jacobianMatrix(1, 0), jacobianMatrix(2, 0));
    Eigen::Vector3d dX_det(jacobianMatrix(0, 1), jacobianMatrix(1, 1), jacobianMatrix(2, 1));

    Eigen::Vector3d normalVector = dX_dxi.cross(dX_det);
    double length = normalVector.norm();

    if (length < 1e-12)
    {
        std::cerr << "Warning: Degenerate triangle: nodes are collinear or coincident. Returning zero matrix." << std::endl;
        return Eigen::MatrixXd::Zero(totalDof, totalDof);
    }

    normalVector.normalize();
    Eigen::Vector3d localZ = normalVector;

    Eigen::Matrix3d globalIdentityMatrix = Eigen::Matrix3d::Identity();
    double dotProductWithGlobalX = normalVector.x();
    Eigen::Vector3d localX;

    if (std::abs(dotProductWithGlobalX) > 1.0 - 1e-10)
    {
        double dotProductWithGlobalZ = normalVector.z();
        localX = globalIdentityMatrix.col(2) - dotProductWithGlobalZ * normalVector;
    }
    else
    {
        localX = globalIdentityMatrix.col(0) - dotProductWithGlobalX * normalVector;
    }

    length = localX.norm();
    if (length < 1e-12)
    {
        std::cerr << "Warning: Degenerate local X-axis detected. Returning zero matrix." << std::endl;
        return Eigen::MatrixXd::Zero(totalDof, totalDof);
    }
    localX.normalize();

    Eigen::Vector3d localY = localZ.cross(localX);
    length = localY.norm();
    if (length < 1e-12)
    {
        std::cerr << "Warning: Degenerate local Y-axis detected. Returning zero matrix." << std::endl;
        return Eigen::MatrixXd::Zero(totalDof, totalDof);
    }
    localY.normalize();

    Eigen::Matrix3d transformationMatrixLocal;
    transformationMatrixLocal.row(0) = localX.transpose();
    transformationMatrixLocal.row(1) = localY.transpose();
    transformationMatrixLocal.row(2) = localZ.transpose();

    Eigen::MatrixXd transformationMatrixGlobal = Eigen::MatrixXd::Zero(totalDof, totalDof);

    for (int nodeIdx = 0; nodeIdx < nodesCount; ++nodeIdx)
    {
        int translationalDofOffset = nodeIdx * dof;
        int rotationalDofOffset = translationalDofOffset + dimensions;

        for (int i = 0; i < dimensions; ++i)
        {
            for (int j = 0; j < dimensions; ++j)
            {
                transformationMatrixGlobal(translationalDofOffset + i, translationalDofOffset + j) = transformationMatrixLocal(i, j);
                transformationMatrixGlobal(rotationalDofOffset + i, rotationalDofOffset + j) = transformationMatrixLocal(i, j);
            }
        }
    }

    return transformationMatrixGlobal;
}

File: hekatan-fem/src/utils/getTransformationMatrix.ts
import { Node } from ".././data-model";
import {
  cross,
  dot,
  identity,
  kron,
  MathCollection,
  Matrix,
  norm,
  subtract,
  zeros,
} from "mathjs";
import { getTransformationMatrixShellQ4 } from "./shellQ4";

// from global to local
export function getTransformationMatrix(nodes: Node[]): number[][] {
  if (nodes.length === 2) return getTransformationMatrixFrame(nodes);
  if (nodes.length === 3) return getTransformationMatrixShell(nodes);
  if (nodes.length === 4) return getTransformationMatrixShellQ4(nodes);
}

/**
 * Ejes locales de barra, convención CSI (SAP2000 / ETABS):
 *
 *   eje 1 (x) = del nudo i al nudo j
 *   eje 2 (y) = en el plano vertical que contiene la barra, hacia ARRIBA.
 *               Si la barra es vertical no hay tal plano: se toma el +X global.
 *   eje 3 (z) = eje1 × eje2  (queda horizontal en barras no verticales)
 *
 * Antes la tríada era otra: y = (-m/D, l/D, 0) horizontal y z hacia arriba.
 * Las dos son dextrógiras y las dos resuelven igual, pero respecto a CSI la
 * vieja giraba 90° EN UN SENTIDO en las vigas y en el CONTRARIO en las
 * columnas (viga: y=-eje3, z=+eje2; columna: y=+eje3, z=-eje2). O sea que no
 * había un mapeo único para leer las fuerzas de barra contra ETABS, y el
 * signo de V y M cambiaba según el tipo de barra. Medido contra
 * FrameObj.GetTransformationMatrix del propio ETABS.
 *
 * CONSECUENCIA EN LOS DATOS DE ENTRADA: al girar los ejes, la inercia que
 * gobierna cada plano cambia de casillero.
 *   momentsOfInertiaZ = I33 (flexión en el plano 1-2: V2 y M3)
 *   momentsOfInertiaY = I22 (flexión en el plano 1-3: V3 y M2)
 * Con la tríada vieja era justo al revés. Los modelos que pasan las dos
 * inercias distintas hay que cruzarlos.
 */
function getTransformationMatrixFrame(nodes: Node[]): number[][] {
  const vector = subtract(nodes[1], nodes[0]) as number[];
  const length = norm(vector) as number;
  const l = dot(vector, [1, 0, 0]) / length;
  const m = dot(vector, [0, 1, 0]) / length;
  const n = dot(vector, [0, 0, 1]) / length;
  const D = Math.sqrt(l ** 2 + m ** 2);

  // Vertical: el plano vertical no define nada, CSI fija el eje 2 en +X global.
  // Se compara con tolerancia, no con ===: un nudo escrito con 6 decimales deja
  // n = 0.9999999999 y caía en la rama general, que con D→0 da infinitos.
  if (D < 1e-9) {
    const s = n > 0 ? 1 : -1;
    const lambda = [
      [0, 0, s],        // eje 1: a lo largo de la barra
      [1, 0, 0],        // eje 2: +X global
      [0, s, 0],        // eje 3 = eje1 × eje2
    ];
    return kron(identity(4) as MathCollection, lambda).toArray() as number[][];
  }

  const lambda = [
    [l, m, n],
    [(-l * n) / D, (-m * n) / D, D],
    [m / D, -l / D, 0],
  ];

  return kron(identity(4) as MathCollection, lambda).toArray() as number[][];
}

function getTransformationMatrixShell(nodes: Node[]): number[][] {
  const nodesCount = 3;
  const dimensions = 3;
  const dof = 6;
  const totalDof = nodesCount * dof;

  const globalCoordinates = [nodes[0], nodes[1], nodes[2]];
  const localCoordinates = (
    zeros(dimensions, nodesCount) as Matrix
  ).toArray() as number[][];

  for (let i = 0; i < dimensions; i++) {
    for (let j = 0; j < nodesCount; j++) {
      localCoordinates[i][j] = globalCoordinates[j][i];
    }
  }

  const dN_dxi = [-1.0, 1.0, 0.0];
  const dN_det = [-1.0, 0.0, 1.0];
  const jacobianMatrix = (
    zeros(dimensions, 2) as Matrix
  ).toArray() as number[][];

  for (let i = 0; i < dimensions; i++) {
    for (let k = 0; k < nodesCount; k++) {
      jacobianMatrix[i][0] += localCoordinates[i][k] * dN_dxi[k];
      jacobianMatrix[i][1] += localCoordinates[i][k] * dN_det[k];
    }
  }

  const dX_dxi = jacobianMatrix.map((row) => row[0]);
  const dX_det = jacobianMatrix.map((row) => row[1]);
  let normalVector = cross(dX_dxi, dX_det) as number[];
  let length = norm(normalVector) as number;

  if (length === 0) {
    console.warn("Degenerate triangle: nodes are collinear or coincident.");
    return (zeros(totalDof, totalDof) as Matrix).toArray() as number[][];
  }

  normalVector = normalVector.map((v) => v / length);
  const localZ = [...normalVector];

  const globalIdentityMatrix = (
    identity(dimensions) as Matrix
  ).toArray() as number[][];
  const dotProductWithGlobalX = normalVector[0];
  let localX: number[];

  if (Math.abs(dotProductWithGlobalX) > 1 - 1e-10) {
    const dotProductWithGlobalZ = normalVector[2];
    localX = globalIdentityMatrix.map(
      (row, l) => row[2] - dotProductWithGlobalZ * normalVector[l]
    );
  } else {
    localX = globalIdentityMatrix.map(
      (row, l) => row[0] - dotProductWithGlobalX * normalVector[l]
    );
  }

  length = norm(localX) as number;
  if (length === 0) {
    console.warn("Degenerate local X-axis detected.");
    return (zeros(totalDof, totalDof) as Matrix).toArray() as number[][];
  }
  localX = localX.map((v) => v / length);

  let localY = cross(localZ, localX) as number[];
  length = norm(localY) as number;
  if (length === 0) {
    console.warn("Degenerate local Y-axis detected.");
    return (zeros(totalDof, totalDof) as Matrix).toArray() as number[][];
  }
  localY = localY.map((v) => v / length);

  const transformationMatrixLocal = [localX, localY, localZ];

  const transformationMatrixGlobal = (
    zeros(totalDof, totalDof) as Matrix
  ).toArray() as number[][];

  for (let nodeIdx = 0; nodeIdx < nodesCount; nodeIdx++) {
    const translationalDofOffset = nodeIdx * dof;
    const rotationalDofOffset = translationalDofOffset + dimensions;

    for (let i = 0; i < dimensions; i++) {
      for (let j = 0; j < dimensions; j++) {
        transformationMatrixGlobal[translationalDofOffset + i][
          translationalDofOffset + j
        ] = transformationMatrixLocal[i][j];
        transformationMatrixGlobal[rotationalDofOffset + i][
          rotationalDofOffset + j
        ] = transformationMatrixLocal[i][j];
      }
    }
  }

  return transformationMatrixGlobal;
}
